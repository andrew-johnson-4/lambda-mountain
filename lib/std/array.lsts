
declare-binop( $"!=", raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l"!="; y; l")"; ) );
declare-binop( $"==", raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l"=="; y; l")"; ) );
declare-binop( $"<",  raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l"<"; y; l")"; ) );
declare-binop( $"<=", raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l"<="; y; l")"; ) );
declare-binop( $">",  raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l">"; y; l")"; ) );
declare-binop( $">=", raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l">="; y; l")"; ) );

declare-binop( $"+", raw-type(base-type[]), raw-type(U64), raw-type(base-type[]), ( l"("; x; l"+"; y; l")"; ) );
declare-binop( $"+", raw-type(base-type[]), raw-type(I64), raw-type(base-type[]), ( l"("; x; l"+"; y; l")"; ) );
declare-binop( $"-", raw-type(base-type[]), raw-type(U64), raw-type(base-type[]), ( l"("; x; l"-"; y; l")"; ) );
declare-binop( $"-", raw-type(base-type[]), raw-type(I64), raw-type(base-type[]), ( l"("; x; l"-"; y; l")"; ) );
declare-binop( $"-", raw-type(base-type[]), raw-type(base-type[]), raw-type(U64), ( l"("; x; l"-"; y; l")"; ) );

declare-unop( not, raw-type(base-type[]), raw-type(U64), ( l"(!"; x; l")"; ) );

declare-unop( into-branch-conditional, raw-type(base-type[]), raw-type(U64), x );

declare-binop( $"[]", raw-type(base-type[]), raw-type(U64), raw-type(base-type), ( l"("; x; l"["; y; l"])"; ) );
declare-ternop( $"set[]", raw-type(base-type[]), raw-type(U64), raw-type(base-type), raw-type(base-type), ( l"("; x; l"["; y; l"]="; z; l")"; ) );

declare-unop( open, raw-type(t), raw-type(t), x );
declare-unop( open, raw-type(base-type[]), raw-type(base-type), (l"(*"; x; l")";) );
declare-unop( $"&", raw-type(t), raw-type(t[]), (l"(&"; x; l")";) );

declare-unop( raw, raw-type(t), raw-type(t), x );
declare-unop( raw, raw-type(base-type[]), raw-type(base-type[]+Raw), x );

let mark-memory-as-safe(ptr: t[], len: U64): Nil = (
   # BEFORE CHANGING THIS: talk to alex

   while len > 0_u64 {
      let ignored = ptr[0_u64];
      len = len - 1_u64;
      ptr = ((ptr as U8[]) + sizeof(t)) as t[];
   };
);

let safe-alloc-impl(nb: U64): ?[] = (
   # BEFORE CHANGING THIS: talk to alex

   let ptr = malloc(nb);
   if ptr as U64 == 0_u64 {
      fail("malloc \{nb}B fail");
   };

   mark-memory-as-safe(ptr as U8[], nb);
   ptr
);

let safe-realloc-impl(ptr: ?[], nb: U64): ?[] = (
   # BEFORE CHANGING THIS: talk to alex

   let new_ptr = realloc(ptr, nb) as ?[];
   if new_ptr as U64 == 0_u64 {
      fail("realloc to \{nb}B fail");
   };

   mark-memory-as-safe(new_ptr as U8[], nb);
   new_ptr
);

## this will fail() if len is 0
let safe-alloc(len: U64, ty: Type<t>): t[] = (
   # BEFORE CHANGING THIS: talk to alex

   let nb = len * sizeof(t);
   safe-alloc-impl(nb) as t[]
);

## this will fail() if len is 0
let safe-realloc(ptr: t[], len: U64, ty: Type<t>): t[] = (
   # BEFORE CHANGING THIS: talk to alex

   let nb = len * sizeof(t);
   safe-realloc-impl(ptr as ?[], nb) as t[]
);

let close(x: p): p[] = (
   let r = safe-alloc(1_u64, type(p));
   r[0_u64] = x;
   r
);
