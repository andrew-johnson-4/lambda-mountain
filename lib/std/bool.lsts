
type alias Bool = U64;

let true = 1_u64;
let false = 0_u64;
