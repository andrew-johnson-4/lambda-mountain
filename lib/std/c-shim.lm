
import stdio.h;
import stdlib.h;
import string.h;
import regex.h;

strcmp := λ: FFI(: l U8[])(: r U8[]). (: __uninitialized I32);
memcmp := λ: FFI(: l ?[])(: r ?[])(: len U64). (: __uninitialized U32);
printf := λ: FFI(: format String)(: arg Any). (: __uninitialized U32);
fprintf := λ: FFI(: io IO::File)(: format String)(: arg Any). (: __uninitialized U32);
putchar := λ: FFI(: x U32). (: __uninitialized U32);
exit := λ: FFI(: x U32). (: __uninitialized Nil);
malloc := λ: FFI(: sz U64). (: __uninitialized ?[]);
realloc := λ: FFI(: ptr ?[])(: sz U64). (: __uninitialized ?[]);
memset :=  λ: FFI(: ptr ?[])(: x U32)(: sz U64). (: __uninitialized ?[]);
fopen := λ: FFI(: fp U8[])(: mode U8[]). (: __uninitialized IO::File);
fread := λ: FFI(: buff ?[])(: size U64)(: count U64)(: f IO::File). (: __uninitialized U64);
fwrite := λ: FFI(: buff ?[])(: size U64)(: count U64)(: f IO::File). (: __uninitialized U64);
fclose := λ: FFI(: f IO::File). (: __uninitialized U32);
execvp := λ: FFI(: file U8[])(: argv U8[][]). (: __uninitialized U32);
mktemp := λ: FFI(: template U8[]). (: __uninitialized U8[]);
system := λ: FFI(: command U8[]). (: __uninitialized U32);
