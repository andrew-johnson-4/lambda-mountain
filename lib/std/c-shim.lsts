
import stdio.h;
import stdlib.h;
import string.h;
import regex.h;

let :FFI strcmp(l: U8[], r: U8[]): I32 = ();
let :FFI memcmp(l: ?[], r: ?[], len: U64): U32 = ();
let :FFI printf(format: CString, arg: Any): U32 = ();
let :FFI fprintf(io: IO::File, format: CString, arg: Any): U32 = ();
let :FFI putchar(x: U32): U32 = ();
let :FFI exit(x: U32): Nil = ();
let :FFI malloc(sz: U64): ?[] = ();
let :FFI realloc(ptr: ?[], sz: U64): ?[] = ();
let :FFI memset(ptr: ?[], x: U32, sz: U64): ?[] = ();
let :FFI fopen(fp: U8[], mode: U8[]): IO::File = ();
let :FFI fread(buff: ?[], size: U64, count: U64, f: IO::File): U64 = ();
let :FFI fwrite(buff: ?[], size: U64, count: U64, f: IO::File): U64 = ();
let :FFI fclose(f: IO::File): U32 = ();
let :FFI execvp(file: U8[], argv: U8[][]): U32 = ();
let :FFI mktemp(template: U8[]): U8[] = ();
let :FFI system(command: U8[]): U32 = ();

