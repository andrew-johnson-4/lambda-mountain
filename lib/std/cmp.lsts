

## order is important here because
## [type(LessThan).discriminator-case-tag] = 0
## [type(Equal).discriminator-case-tag] = 1
## [type(GreaterThan).discriminator-case-tag] = 2
type Ord = LessThan | Equal | GreaterThan;

let $"=="(l: Ord, r: Ord): U64 = l.discriminator-case-tag == r.discriminator-case-tag;
let $"!="(l: Ord, r: Ord): U64 = l.discriminator-case-tag != r.discriminator-case-tag;
let $"<"(l: Ord, r: Ord): U64  = l.discriminator-case-tag <  r.discriminator-case-tag;
let $"<="(l: Ord, r: Ord): U64 = l.discriminator-case-tag <= r.discriminator-case-tag;
let $">"(l: Ord, r: Ord): U64  = l.discriminator-case-tag >  r.discriminator-case-tag;
let $">="(l: Ord, r: Ord): U64 = l.discriminator-case-tag >= r.discriminator-case-tag;

let $"=="(l: x, r: x): U64 = cmp(l, r) == Equal();
let $"!="(l: x, r: x): U64 = cmp(l, r) != Equal();
let $"<"(l: x, r: x): U64  = cmp(l, r) <  Equal();
let $"<="(l: x, r: x): U64 = cmp(l, r) <= Equal();
let $">"(l: x, r: x): U64  = cmp(l, r) >  Equal();
let $">="(l: x, r: x): U64 = cmp(l, r) >= Equal();

let $"&&"(l: Ord, r: Ord): Ord = if l == Equal() then r else l;

#TODO print(IO::File, String)
#let print(io: IO::File, o: Ord): Nil = (
#   if o == GreaterThan then print(io, "GreaterThan") else
#   if o == Equal then print(io, "Equal") else
#   print(io, "LessThan");
#);
