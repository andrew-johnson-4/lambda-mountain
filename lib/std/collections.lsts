
interface self implements Collection<t> {
   let .length(x: self): U64;
   let $"[]"(x: self, idx: U64): t;
}

let .last(self: t+Collection<x>): x = (
   let ln = self.length();
   if ln == 0 {
      fail("cannot get .last() from empty collection");
   };
   self[(ln - 1)]
);

interface self implements Iterator<t> {
   let .next(x: self[]): Maybe<t>;
}

interface self implements Iterable<t> {
   # TODO: uncomment doc
   # ## When modifying th eunderlying collection while iterating,
   # ## the iterator should continue on the old version.
   let .iter(x: self): Iterator<t>;

   # ## Underlying collection not allowed to be modified while iterating.
   let .unsafe-iter(x: self): Iterator<t>;
}

type Iterator<t> implements Iterable<t>;
let .iter(x: t+Iterator<elt>): t = ( x );
let .unsafe-iter(x: t+Iterator<elt>): t = ( x );

interface IteratorRemaining {
   let .remaining(x: self + Iterator<t>): U64;
};

let remaining-or-zero(from: Iterator<x>): U64 = (
   0
);

let remaining-or-zero(from: Iterator<x>+IteratorRemaining): U64 = (
   from.remaining()
);

let length-or-zero(from: Iterable<x>): U64 = (
   0_u64
);

let length-or-zero(from: Collection<x>): U64 = (
   from.length()
);

let length-or-zero(from: Collection<x>+Iterable<x>): U64 = (
   from.length()
);

## TODO: remove
type Unused<t> = UnusedType{};

type CollectionIteratorBase<c,t> = { coll: c, idx: U64, _u: Unused<t> };
type CollectionIteratorBase<c,t> implements Iterator<t>;
type CollectionIteratorBase<c,t> implements IteratorRemaining;

let .remaining(self: CollectionIteratorBase<c,t>): U64 = (
   self.coll.length() - self.idx
);

let .next(p: CollectionIteratorBase<c,t>[]): Maybe<t> = (
   let v = p[0];
   if v.idx < v.coll.length() {
      let val = v.coll[v.idx];
      v.idx = v.idx + 1;
      p[0] = v;
      Some(val) : Maybe<t>
   } else {
      None : Maybe<t>
   }
);

let .into(self: c+Collection<t>, out: Type<CollectionIteratorBase<c,t>>): CollectionIteratorBase<c,t> = (
   CollectionIteratorBase( self, 0_u64, UnusedType() : Unused<t> )
);
