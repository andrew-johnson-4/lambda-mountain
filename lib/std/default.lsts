
import core/index.lsts;
import std/std-primitives.lsts;
import std/c-shim.lsts;
import std/std-macros.lsts;
import std/platform-macros.lsts;
import std/primitives.lsts;
import std/field-accessors.lsts;
import std/sized.lsts;
import std/bool.lsts;
import std/u8.lsts;
import std/i8.lsts;
import std/u16.lsts;
import std/i16.lsts;
import std/u32.lsts;
import std/i32.lsts;
import std/u64.lsts;
import std/i64.lsts;
import std/f64.lsts;
import std/usize.lsts;
import std/string.lsts;
import std/smart-string.lsts;
import std/s.lsts;
import std/tuple.lsts;
import std/list.lsts;
import std/maybe.lsts;
import std/vector.lsts;
import std/hashtable.lsts;
import std/array.lsts;
import std/io.lsts;
import std/cmp.lsts;
import std/print.lsts;
import std/collections.lsts;
import std/regex.lsts;
