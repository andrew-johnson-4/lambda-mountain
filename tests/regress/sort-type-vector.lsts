
import std/default.lsts;

if A_s < B_s then print("A<B\n") else print("B<A\n"); 
if B_s < A_s then print("B<A\n") else print("A<B\n"); 

if BA_s < B_s then print("BA<B\n") else print("B<BA\n"); 
if B_s < BA_s then print("B<BA\n") else print("BA<B\n"); 

if "A" == "A" then print("A==A\n") else print("A!=A\n"); 
if "A" == "B" then print("A==B\n") else print("A!=B\n"); 

if "A" < "B" then print("A<B\n") else print("A>=B\n"); 
if "B" < "A" then print("B<A\n") else print("B>=A\n"); 

if "BA" < "B" then print("BA<B\n") else print("BA>=B\n"); 
if "B" < "BA" then print("B<BA\n") else print("B>=BA\n"); 

# TODO: Restore this functionality as a promise
#if ["A"] < ["A", "B"] then print("[A]<[A,B]\n") else print("[A]>=[A,B]\n");
#if ["A"] < ["A"] then print("[A]<[A]\n") else print("[A]>=[A]\n");
#if ["A"] < ["B"] then print("[A]<[B]\n") else print("[A]>=[B]\n");
#if ["B"] < ["A"] then print("[B]<[A]\n") else print("[B]>=[A]\n");
