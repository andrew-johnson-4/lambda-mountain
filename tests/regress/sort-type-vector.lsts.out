A<B
A<B
B<BA
B<BA
A==A
A!=B
A<B
B>=A
BA>=B
B<BA
